[pvr] - rework PVRManager, addons are owned by addon system

FernetMenta opened the pull request:
Addons are not owned by PVR, just managed. PVRManager is required if at least one PVR addon is enabled. Explicitly enable/disable PVRManager makes no sense anymore and is gone.

@ksooo @Jalle19 @xhaggi

FernetMenta committed:
[pvr] - rework PVRManager, addons are owned by addon system

Jalle19 commented on xbmc/pvr/PVRManager.cpp:
```diff
@@ -92,12 +92,6 @@ const int CPVRManager::m_pvrWindowIds[12] = {
 
 CPVRManager::CPVRManager(void) :
     CThread("PVRManager"),
-    m_channelGroups(NULL),
```
Why drop these?

FernetMenta replied:
those are unique_ptr now, no need to init twice

Jalle19 commented on xbmc/pvr/PVRManager.h:
```diff
@@ -666,11 +666,11 @@ namespace PVR
-    CPVRTimers *                    m_timers;                      /*!< pointer to the timers container */
-    CPVRClients *                   m_addons;                      /*!< pointer to the pvr addon container */
-    CPVRGUIInfo *                   m_guiInfo;                     /*!< pointer to the guiinfo data */
+    std::unique_ptr<CPVRChannelGroupsContainer>    m_channelGroups;               /*!< pointer to the channel groups container */
```
Nevermind my previous comment, this explains it

phil65 commented:
@FernetMenta is this here still needed then?
https://github.com/phil65/xbmc/blob/master/xbmc/pvr/addons/PVRClients.cpp#L1392

Jalle19 commented:
Seems like a good start on getting multiple simultaneous clients to actually work. It's should be easier now that we can assume the PVR manager id always there and "enabled".

FernetMenta commented:
@phil65 nope, it's removed

@Jalle19 do you mean multiple clients of the same kind or different?

phil65 commented:
Oh sorry, missed to spot that in the diff.

ksooo commented on xbmc/addons/PVRClient.cpp:
```diff
@@ -126,7 +122,7 @@ void CPVRClient::OnPreInstall()
 void CPVRClient::OnPostInstall(bool update, bool modal)
 {
   // (re)start the pvr manager
```
maybe you should remove this comment?

ksooo commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -109,9 +104,14 @@ int CPVRClients::GetClientId(const AddonPtr client) const
+  {
+    if (entry.second->ID() == client->ID())
+    {
+      int sdf = entry.first;
```
sdf is unused

ksooo commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -1141,180 +1135,49 @@ bool CPVRClients::IsKnownClient(const AddonPtr client) const
-          }
-        }
+        PVR_CLIENT pvrclient = std::dynamic_pointer_cast<CPVRClient>(addon);
+        pvrclient.get()->Create(iClientId);
```
not checking the return value of a dynamic_cast imo causes a coverity issue

ksooo commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -1389,72 +1252,26 @@ bool CPVRClients::AutoconfigureClients(void)
-  if (!g_PVRManager.IsStarted() && !g_PVRManager.IsInitialising())
-    return;
-
-  CGUIDialogOK::ShowAndGetInput(CVariant{19240}, CVariant{19241});
```
have you checked whether one or both of the string resources are now unused?

ksooo commented on xbmc/pvr/addons/PVRClients.h:
```diff
@@ -64,9 +64,9 @@ namespace PVR
-                      private CThread
+                      public Observer
   {
+  friend class CPVRClient;
```
why do we need this now?

FernetMenta replied:
because CPVRClient::OnEnable/Disable call CPVRClients::UpdateAddons

ksooo replied:
... and you prefer a <code>friend</code> to be able to call one single method over making that method <code>public</code>?

FernetMenta replied:
```
  #define g_PVRManager       CPVRManager::GetInstance()
  #define g_PVRChannelGroups g_PVRManager.ChannelGroups()
  #define g_PVRTimers        g_PVRManager.Timers()
  #define g_PVRRecordings    g_PVRManager.Recordings()
  #define g_PVRClients       g_PVRManager.Clients()
```

yes I do when dealing with this. at least fo the time being. those public methods gets called from everywhere

ksooo replied:
fair enough

ksooo commented on xbmc/pvr/windows/GUIWindowPVRBase.cpp:
```diff
@@ -359,7 +359,7 @@ void CGUIWindowPVRBase::SetInvalid()
 bool CGUIWindowPVRBase::CanBeActivated() const
 {
   // No activation if PVR is not enabled.
-  return CSettings::GetInstance().GetBool(CSettings::SETTING_PVRMANAGER_ENABLED);
```
This was introduced to disable to open empty (ugly, maybe not working) pvr windows by for instance accidentally hitting the respective keyboard shortcut. This will not work anymore, now.

Either we fix the implementation or remove CGUIWindowPVRBase::CanBeActivated.

I opt for the first. Imo <code>return g_PVRClients->HasCreatedClients()</code> should do the trick.

FernetMenta replied:
it should have tested if PVRManager was running, not if it was enabled. On my list ...

ksooo commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -1141,180 +1135,49 @@ bool CPVRClients::IsKnownClient(const AddonPtr client) const
-      disableAddons.push_back(*it);
+  for (auto &addon : map)
+  {
+    bool bEnabled = !CAddonMgr::GetInstance().IsAddonDisabled((addon)->ID());
```
are the brackets around <code>addon</code> really needed?

FernetMenta replied:
probably not :)

Jalle19 commented:
@FernetMenta different

FernetMenta committed:
[pvr] do not access AddonDatabase for clientID

FernetMenta committed:
[pvr] - drop AutoConfigure, we cant hijack the thread of the event

FernetMenta commented:
@Jalle19 why did this not work?

FernetMenta commented:
@ksooo updated according to your comments.

- I dropped access to AddonsDatabase. PVR used addon's primary key as clientID :( I changed this to a hash
- Dropped AutoConfigure. It did not fit anymore and we can't hijack the thread it was running on.

FernetMenta commented:
@stefansaraev could you test again please

Jalle19 commented:
@FernetMenta I think it's because the PVR manager doesn't get properly notified when an addon is enabled, so it doesn't load any channels etc. from it. Probably worth a fresh look after this gets in.

Jalle19 commented:
@FernetMenta ref: https://github.com/xbmc/xbmc/pull/7020

FernetMenta commented:
@Jalle19 this code restarts/starts/stops PVRManager if PVR addons get enabled/disabled

@ksooo what collisions?

ksooo commented:
@FernetMenta if hash function returns let's say 1 for client 1 and -1 for client 2, then resulting client id for both clients would be 1. Or is it guaranteed that this cannot happen? Don't think so.

FernetMenta committed:
[pvr] - drop needless calls to PVRManager.IsStarted in VideoPlayer

FernetMenta committed:
[pvr] - addon, rework overrides

FernetMenta commented:
@AchimTuran ADSP has copied a lot from PVR and needs rework too.

stefansaraev commented:
@FernetMenta now better.

FernetMenta committed:
[pvr] - no need to observe AddonManager

FernetMenta committed:
[pvr] - reconnect backend after settings have been changed

da-anda commented:
It's been requested in the forum several times to get the possibility to have multiple instances of the same PVR backend running (without having to copy and rename the add-on folder). I know that this will require additional changes (like a connection manager on which you can select the backend to use and apply specific config for this connection), but it would be good to keep this scenario in mind when reworking things.

da-anda commented:
sorry for the probably stupid question, but shouldn't `CApplication::StopPVRManager()` also be removed and instead be handled in `CServiceManager::Deinit()` ?
Also, what happens in Channel-Manager etc which settings are now no longer safe guarded to require a working PVR setup? Instead of dropping the settings condition, shouldn't we replace it with a different one that makes sure that we actually have channels to manage etc? What's the opinion of our PVR devs on this?

da-anda commented on xbmc/addons/PVRClient.cpp:
```diff
@@ -94,61 +94,40 @@ CPVRClient::~CPVRClient(void)
-  if (CPVRManager::GetInstance().IsStarted() && CPVRManager::GetInstance().RestartManagerOnAddonDisabled())
-    CPVRManager::GetInstance().Start(true);
+  CAddon::OnDisabled();
+  CPVRManager::GetInstance().Clients()->UpdateAddons();
```
in order to get rid of singletons, shouldn't you ask the ServiceManager for the PVR instance instead? Or is this change something for later?

FernetMenta replied:
yes, but I wanted to leave some work for the others :)

FernetMenta commented:
@da-anda before a use could enable PVR and then click on channel manager. There was no check behind. So actually nothing has changed.

FernetMenta commented:
jenkins build this please

FernetMenta commented:
> sorry for the probably stupid question, but shouldn't CApplication::StopPVRManager() also be removed and instead be handled in CServiceManager::Deinit() ?

@da-anda maybe next round because this requires more rework. Currently it would need kind of Deinit1/Dinit2

FernetMenta commented:
If no objections, I would merge this.
This is an important step to splitting off streaming relevant functions into the inputstream extension point. Then Kodi will be able to open multiple streams on a backend that enables thumbnail support for PVR and finally PIP.

ksooo commented:
+1 go ahead. Thx.

AchimTuran commented:
@FernetMenta @AlwinEsch 
> @AchimTuran ADSP has copied a lot from PVR and needs rework too.

Well where should I start. AudioDSP has a Singleton, a thread which is polling for addon updates/enable/disable, the processing is done inside bufferpool and some parts are really spaghetti code :-(

But hehe it's the first iteration and quite a complex system. Sometimes it works well and sometimes not. I use it daily for stereo upmixing with adsp.freesurround, which sounds much better than ProLogic II :-D
My plan is to release my reworked [adsp.template](https://github.com/AchimTuran/adsp.template) and adapt all available add-ons to this "small" framework. Afterwards it is required to support all platforms from core- and the add-on-side. It's a long road and any contribution is welcome.

FernetMenta commented:
@AchimTuran study this PR. All this addon handling needs to be changed. ADSP is not the owner of the addons.

simonbuehler commented:
@FernetMenta i used to call PVR.Stop/Startmanager from a Idle-Addon so that the connection to my vdr server gets dropped when kodi is idle and the vdr can handle the logic of suspend/hibernate when no clients are connected.
As this functionality is gone, how can i force kodi to disconnect from the vdr server now?

FernetMenta commented:
you have to disable the addon. afaik this is possible with json-rpc
